cli/replicaset/cconfig.go

```diff
@@ -204,3 +218,408 @@ func updateCConfigInstances(replicaset *Replicaset, topology cconfigTopology) {
 		}
 	}
 }
+
+// Promote promotes an instance in the application.
+func (c *CConfigApplication) Promote(ctx PromoteCtx) error {
+	replicasets, err := c.GetReplicasets()
+	if err != nil {
+		return fmt.Errorf("failed to get replicasets: %w", err)
+	}
+
+	var (
+		targetReplicaset Replicaset
+		targetInstance   Instance
+		found            bool
+	)
+loop:
+	for _, replicaset := range replicasets.Replicasets {
+		for _, instance := range replicaset.Instances {
+			if instance.Alias == ctx.InstName {
+				targetReplicaset = replicaset
+				targetInstance = instance
+				found = true
+				break loop
+			}
+		}
+	}
+	if !found {
+		return fmt.Errorf("instance %q not found in a configured replicaset", ctx.InstName)
+	}
+	if !targetInstance.InstanceCtxFound {
+		return fmt.Errorf("instance %q should be online", ctx.InstName)
+	}
+
+	var instances []running.InstanceCtx
+	var unfound []string
+	for _, inst := range targetReplicaset.Instances {
+		if !inst.InstanceCtxFound {
+			unfound = append(unfound, inst.Alias)
+		} else {
+			instances = append(instances, inst.InstanceCtx)
+		}
+	}
+	if len(unfound) > 0 && !ctx.SkipUnfoundInstances {
+		return fmt.Errorf("all instances in the target replicaset should "+
+			"be online, could not connect to: %s", strings.Join(unfound, ","))
+	}
+
+	isConfigPatched, err := cconfigPromote(targetInstance.InstanceCtx, ctx)
+	if err != nil {
+		return err
+	}
+	if isConfigPatched {
+		err = reloadCConfig(instances)
+	}
+	return err
+}
+
+// cconfigPromoteElection tries to promote an instance via `box.ctl.promote()`.
+func cconfigPromoteElection(evaler connector.Evaler, timeout int) error {
+	args := []any{timeout}
+	opts := connector.RequestOpts{}
+	_, err := evaler.Eval(cconfigPromoteElectionBody, args, opts)
+	return err
+}
+
+// CConfigPublisher publishes the config.
+type CConfigPublisher interface {
+	// Publish publishes the config into a specific key that contains the config prefix.
+	Publish(key string, data []byte, revision int64) error
+}
+
+// CConfigSource is a cluster config source.
+type CConfigSource struct {
+	dataCollector libcluster.DataCollector
+	publisher     CConfigPublisher
+}
+
+// NewConfigSource creates CConfigSource.
+func NewCConfigSource(dataCollector libcluster.DataCollector,
+	publisher CConfigPublisher) CConfigSource {
+	return CConfigSource{
+		dataCollector: dataCollector,
+		publisher:     publisher,
+	}
+}
+
+// cconfigDataCollector collects and caches the config data.
+type cconfigDataCollector struct {
+	rawCollector libcluster.DataCollector
+	cached       bool
+	data         []libcluster.Data
+}
```

Comment:
* Want to get `[]libcluster.Data` and also merged data. 
* Don't want to write merging logic myself, because it's libcluster responsibility, how to merge configs by the different keys.
* Can't pass `[]libcluster.Data` to the libcluster to merge it inside. (It's ok, just observation). https://github.com/tarantool/tt/blob/master/lib/cluster/yaml.go#L41-L55
* So let's create the type, which will cache `[]libcluster.Data` and after libcluster merging we can take the cached `[]libcluster.Data`.

Reply:
It looks like a some general use case. I suggest to patch the `libcluster`.

You could:
1. Patch the YamlCollector to receive vararg of data and merge it into a Config.
https://github.com/tarantool/tt/blob/12ca1406169dfd14dc77a74d5d85e8a836d5d791/lib/cluster/yaml.go#L15-L30
NewYamlCollector(data []byte...)
2. Make a separate type to merge libcluster.Data into a Config (I like that way):
`NewYamlDataMergeCollector(data libcluster.Data...)` or `NewYamlDataMergeCollector(data []libcluster.Data)`

Reply:
As a result, you will just collect []libcluster.Data with a general collector and could work with it without helper subtypes.